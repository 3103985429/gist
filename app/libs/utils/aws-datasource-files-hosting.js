const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
