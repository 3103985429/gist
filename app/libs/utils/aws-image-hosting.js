const request = require('request');
var winston = require('winston');
var aws = require('aws-sdk');
var sharp = require('sharp');
var async = require('async');
const _ = require('lodash');

const { VIEW_TYPE_OBJECT_DETAILS, VIEW_TYPE_GALLERY, VIEW_TYPE_TIMELINE } = require('../../config/views.config');

const BUCKET = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const S3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
    signatureVersion: 'v4',
    region: 'nyc3',
});

module.exports = {
    uploadToS3,
    signedUrlForPutObject,
    getAllIconsForTeam,
    hostImageLocatedAtRemoteURL,
    copyPageAssets,
    copyDatasetAsset,
    getAssetByKey,
};

function uploadToS3(key, response, callback) {

    var payload = {
        Bucket: BUCKET,
        Key: key,
        Body: response,
        ACL: 'public-read',
    };

    S3.upload(payload, function (err, data) {
        if (err) {
            winston.error('AWS S3 write stream error');
            return callback(err);
        }
        return callback(null, data);
    });
}

function _appendKeyToBucket(key) {
    return 'https://' + BUCKET + '.' + process.env.DO_S3_ENDPOINT + '/' + key;
}

function getImageFormat(headers) {
    const imageFormat = _.get(headers, 'content-type', '').split(';')[0].split('/')[1];

    return imageFormat === 'jpg' ? 'jpeg' : imageFormat;
}

function getImageAtUrl(remoteImageSourceURL, callback) {
    var options = {
        url: remoteImageSourceURL,
        encoding: null,
        timeout: 10000,
    };

    request.get(options, function (err, response, data) {
        if ((err && ['ENOTFOUND', 'ETIMEDOUT'].includes(err.code)) || !response) {
            winston.info(`returning url as null, since Could not read the remote image ${remoteImageSourceURL}: ${err}`);
            return callback(null, null);
        }

        if (err) {
            return callback(err, null);
        }

        const imageFormat = getImageFormat(response.headers);

        if (_.get(sharp, `format.${imageFormat}`)) {
            return callback(null, data);
        }

        return callback(null, null);
    });
}

function signedUrlForPutObject(key, fileType, callback) {
    if (fileType.indexOf('svg') >= 0) { //fix space when uploading svg error
        fileType = 'image/svg+xml';
    }

    var params = {
        Bucket: BUCKET,
        Key: key,
        ACL: 'public-read',
        ContentType: fileType,
    };

    S3.getSignedUrl('putObject', params, function (err, signedUrl) {
        callback(err, { putSignedUrl: signedUrl, publicUrl: _appendKeyToBucket(key) });
    });
}

function getAllIconsForTeam(teamSubdomain, callback) {
    S3.listObjects({
        Bucket: BUCKET,
        Prefix: teamSubdomain + '/assets/icon/',
        Marker: teamSubdomain + '/assets/icon/',
    }, function (err, data) {
        if (err) {
            callback(err);
        } else {
            var listOfUrls = [];
            var objects = data.Contents;

            for (var i = 0; i < objects.length; i++) {
                var objKey = objects[i].Key;

                var url = _appendKeyToBucket(objKey);

                listOfUrls.push(url);
            }
            callback(null, listOfUrls);
        }
    });

}

async function _proceedToStreamToHost(imageUrl, resize, folder, docPKey, callback) {
    try {
        const image = sharp(imageUrl);
        const metadata = await image.metadata();

        if (!metadata) {
            winston.info('returning url as null, since Could not read the remote image metadata');
            return callback();
        }

        const { width, hasAlpha } = metadata;
        const size = width < resize.size ? width : resize.size;
        const whiteColorRGB = { r: 255, g: 255, b: 255 };
        const background = hasAlpha ? { ...whiteColorRGB, alpha: 1 } : whiteColorRGB;
        const resizeConfig = { fit: 'contain', background };
        const resizedImage = resize.cropped ? image.resize(size, size, resizeConfig) : image.resize(size);
        const bufferedImage = await resizedImage.toBuffer();
        const keyName = `${folder}${resize.view}/${docPKey}`;

        uploadToS3(keyName, bufferedImage, callback);
    } catch (error) {
        winston.error('has err: ' + error);
        callback(error);
    }
}

function hostImageLocatedAtRemoteURL(folder, remoteImageSourceURL, overwrite, destinationFilenameSansExt, callback) {
    const recommendedSizes = [
        { size: 763, cropped: false, view: VIEW_TYPE_OBJECT_DETAILS },
        { size: 280, cropped: false, view: VIEW_TYPE_GALLERY },
        { size: 155, cropped: true, view: VIEW_TYPE_TIMELINE },
    ];

    getImageAtUrl(remoteImageSourceURL, function (err, imageBuffer) {
        if (err || !imageBuffer) {
            return callback(err);
        } else {
            async.each(recommendedSizes, function (size, cb) {
                if (overwrite) {
                    _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                } else {
                    const keyName = `${folder}${size.view}/${destinationFilenameSansExt}`;
                    const params = {
                        Bucket: BUCKET,
                        Key: keyName,
                    };
                    try {
                        S3.headObject(params, function (err, data) {
                            if (err || !data) {
                                winston.error('Head object err: ' + err);
                            } else {
                                winston.info(' File already uploaded and overwrite is false for object with pKey ' + destinationFilenameSansExt);
                                cb();
                            }
                        });
                    } catch (err) {

                        if (err.code === 'ENOTFOUND' || err.code === 'ETIMEDOUT') {
                            _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                        } else {
                            winston.error(err);
                            cb(err);
                        }
                    }
                }
            }, callback);
        }
    });

}

function copyPageAssets(args, callback) {
    var params = {
        Bucket: BUCKET,
        CopySource: BUCKET + '/' + args.subdomain + '/websites/' + args.originalWebsiteId + '/pages/' + args.originalPageId + '/' + args.imageType + '/' + args.fileName,
        Key: args.subdomain + '/websites/' + args.websiteId + '/pages/' + args.pageId + '/' + args.imageType + '/' + args.fileName,
        ACL: 'public-read',
    };

    S3.copyObject(params, function (err) {
        if (err) {
            winston.error('\' error copying page asset with params: \'', params);
            return callback(err);
        } else {
            callback();
        }
    });
}

async function copyDatasetAsset({ fromSubdomain, toSubdomain, fromId, toId, source, fileName }, callback) {
    const getUrl = ({ id, BUCKET, subdomain = fromSubdomain }) =>
        `${BUCKET ? `${BUCKET}/` : ''}${subdomain}/datasets/${id}/${source}/${fileName}`;
    const copySource = getUrl({ id: fromId, BUCKET });
    const key = getUrl({ id: toId, subdomain: toSubdomain });

    const params = {
        Bucket: BUCKET,
        CopySource: copySource,
        Key: key,
        ACL: 'public-read',
    };

    try {
        await S3.copyObject(params).promise();

        callback(null);
    } catch (e) {
        winston.error(`S3: error copying dataset asset: ${e}`);

        callback(e);
    }
}

function getAssetByKey(key, callback) {
    const params = {
        Bucket: BUCKET,
        Key: key,
    };

    S3.getObject(params, function (err, data) {
        if (err) {
            return callback(err);
        } else {
            return callback(null, data);
        }
    });
}
