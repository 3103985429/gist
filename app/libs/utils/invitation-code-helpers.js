const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
