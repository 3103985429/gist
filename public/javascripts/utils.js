((arrays, Qs, _, callPhantom) => {
    Object.assign(arrays, {
        changeRoutePath,
        updateQuery,
        getDetailsViewUrl,
        getApiRoute,
        getDefaultOptions,
        takeScreenshot,
        equals,
        changeImageColor,
        addResizeEventListener,
        isMobileDevice,
        isMobileBreakpoint,
        isPreviewAndMobile,
        isEmbed,
        getParentHeight,
        getIframeHeight,
        isIE,
    });

    const STAGING_ENV = 'staging';
    const PRODUCTION_ENV = 'production';
    const DEVELOPMENT_ENV = 'development';
    const ENTERPRISE_ENV = 'enterprise';
    let windowWidth;

    function changeRoutePath(uid, vizName, filters = {}, toOmit = []) {
        const [baseUrl, baseParams] = location.href.split('?');
        const isShared = baseUrl.match(/\/s\/[\da-f]+/);
        const url = isShared ?
            baseUrl.replace(/\/s\/([^/]+$)/gi, `/${uid}/${vizName}`) :
            baseUrl.replace(/([^/]+$)/gi, vizName);
        const params = _.flow([Qs.parse, data => _.omit(data, toOmit), data => _.assign(data, filters), Qs.stringify])(
            baseParams
        );
        const paramsWithJoiner = params ? `?${params}` : '';

        return `${url}${paramsWithJoiner}`;
    }

    function updateQuery(options, option, value) {
        const { routePath, routeParams } = arrays.getRouteParams(location.href);
        const { routePath: routePathWithoutFilter, routeParams: routeParamsWithoutFilter } = arrays.getRouteParams(
            options.routePath_withoutFilter
        );

        routeParams[option] = value;
        routeParamsWithoutFilter[option] = value;
        options[option] = value;

        const uri = `${routePath}?${Qs.stringify(routeParams)}`;

        options.routePath_withoutFilter = `${routePathWithoutFilter}?${Qs.stringify(routeParamsWithoutFilter)}`;
        window.history.replaceState({}, '', uri);
    }

    function getDetailsViewUrl(options, id, objectIndex, viewType) {
        const routePathBase = `/${options.array_source_key}/${id}`;

        options.objectIndex = objectIndex;
        options.viewType = viewType;

        return arrays.constructedRoutePath(routePathBase, options.filterObj, options);
    }

    function getParsedSearch(isExternalAccess, subdomain) {
        const locationSearch = _.get(window, 'location.search', '');

        if (isExternalAccess) {
            return _.cond([
                [_.constant(_.isEmpty(subdomain) && _.isEmpty(locationSearch)), _.constant('')],
                [_.constant(_.isEmpty(subdomain)), _.constant(locationSearch)],
                [_.constant(_.isEmpty(locationSearch)), _.constant('?subdomain=' + subdomain)],
                [_.stubTrue, _.constant(`${locationSearch}&subdomain=${subdomain}`)],
            ])();
        }

        return locationSearch;
    }

    function getApiRoute(options, url, isObjectDetails) {
        const { environment, isExternalAccess, subdomain } = options;

        const parsedSearch = getParsedSearch(isExternalAccess, subdomain);

        const PROD_URL = `https://${subdomain}.gist.info`;
        const STAGING_URL = `https://${subdomain}.gist.info`;
        const DEV_URL = `http://${subdomain}.local.arrays.co:9080`;

        const requestUrl = _.cond([
            [equals(STAGING_ENV), _.constant(STAGING_URL)],
            [equals(PRODUCTION_ENV), _.constant(PROD_URL)],
            [equals(ENTERPRISE_ENV), _.constant(PROD_URL)],
            [equals(DEVELOPMENT_ENV), _.constant(DEV_URL)],
            [_.stubTrue, _.constant(DEV_URL)],
        ])(environment);

        if (options.sharedPage && !isObjectDetails) {
            return `${requestUrl}/json-api/v1/sharedpages/${options.sharedPageId}/${parsedSearch}`;
        }

        return `${requestUrl}/json-api/v1/datasources/${options.array_source_key}${url}${parsedSearch}`;
    }

    function getDefaultOptions() {
        const { defaultInstanceId, viewOptions = {} } = arrays;

        return viewOptions[defaultInstanceId] || window.options;
    }

    function takeScreenshot(timeout) {
        if (!_.isFunction(callPhantom)) {
            return;
        }

        const takeShot = () => callPhantom('takeShot');

        if (timeout) {
            return setTimeout(takeShot, timeout);
        }

        takeShot();
    }

    function resizeFunc(resizeHandler, allDirection = false) {
        if (allDirection || window.innerWidth !== windowWidth) {
            windowWidth = window.innerWidth;

            resizeHandler();
        }
    }

    function addResizeEventListener(resizeHandler, allDirection) {
        windowWidth = window.innerWidth;

        window.addEventListener(
            'resize',
            _.debounce(() => resizeFunc(resizeHandler, allDirection), 500)
        );
    }

    function equals(a) {
        return b => _.eq(a, b);
    }

    /**
     * Change color of provided image to match selected color.
     * @param {Image} image
     * @param {String} color
     * @param {Function} callback
     */
    function changeImageColor(image, color, callback) {
        const { width, height } = image;
        const canvas = document.createElement('canvas');
        canvas.width = width;
        canvas.height = height;

        const context = canvas.getContext('2d');

        const buffer = document.createElement('canvas');
        buffer.width = width;
        buffer.height = height;
        const bufferContext = buffer.getContext('2d');

        bufferContext.fillStyle = color;
        bufferContext.fillRect(0, 0, width, height);

        bufferContext.globalCompositeOperation = 'destination-atop';
        bufferContext.drawImage(image, 0, 0);

        context.drawImage(image, 0, 0);

        context.globalAlpha = 1;
        context.drawImage(buffer, 0, 0);

        const resultImage = new Image(canvas.width, canvas.height);
        resultImage.onload = () => callback(resultImage);
        resultImage.src = canvas.toDataURL('image/png');
    }

    function isMobileDevice() {
        return typeof window.orientation !== 'undefined' || navigator.userAgent.indexOf('IEMobile') !== -1;
    }

    function isPreviewAndMobile() {
        return arrays.isMobileDevice() && $('.gist-site-content.preview').length;
    }

    function isIE() {
        const { userAgent } = navigator;

        return userAgent.includes('MSIE ') || userAgent.includes('Trident/');
    }

    function isMobileBreakpoint() {
        return window.innerWidth < arrays.constants.CUSTOM_SM_BREAKPOINT;
    }

    function isEmbed(options = {}) {
        return options.embed === 'true' || options.embed === true;
    }

    function getParentHeight() {
        try {
            return window.parent.innerHeight;
        } catch (e) {
            return window.innerHeight;
        }
    }

    function getIframeHeight() {
        const MIN_HEIGHT = 380;
        const MAX_HEIGHT = 780;
        const HEIGHT_RATIO = 0.9;
        const SCREEN_RATIO = 0.8;
        const DEVICE_HEIGHT_RATIO = 0.75;

        const { screen = {} } = window;
        const windowHeight = getParentHeight();

        if (isMobileDevice()) {
            const height = screen.height ? screen.height * SCREEN_RATIO : windowHeight;

            return Math.min(MAX_HEIGHT, Math.round(height * DEVICE_HEIGHT_RATIO));
        }

        return Math.max(MIN_HEIGHT, Math.min(MAX_HEIGHT, windowHeight * HEIGHT_RATIO));
    }
})(window.arrays, window.Qs, window._, window.callPhantom);
