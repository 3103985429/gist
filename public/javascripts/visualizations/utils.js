((arrays, $, _, Qs) => {
    /**
     * Clone (deep copy) a plain Object, Array, Date, String, Number, or Boolean.
     * will work adequately for the 6 simple types I mentioned, as long as the data in the
     * objects and arrays form a tree structure. That is, there isn't more than one reference
     * to the same data in the object.
     *
     * Source: https://stackoverflow.com/questions/728360/how-do-i-correctly-clone-a-javascript-object
     *
     * @param  {Object|Array|Date|String|Number|Boolean} obj - the object to clone
     * @return {Object|Array|Date|String|Number|Boolean}     - a clone (deep copy) of the object
     */
    arrays.clone = function (obj) {
        var copy;

        // Handle the 3 simple types, and null or undefined
        if (obj == null || typeof obj !== 'object') {
            return obj;
        }

        // Handle Date
        if (obj instanceof Date) {
            copy = new Date();
            copy.setTime(obj.getTime());
            return copy;
        }

        // Handle Array
        if (obj instanceof Array) {
            copy = [];
            for (var i = 0, len = obj.length; i < len; i++) {
                copy[i] = arrays.clone(obj[i]);
            }
            return copy;
        }

        // Handle Object
        if (obj instanceof Object) {
            copy = {};
            for (var attr in obj) {
                if (obj.hasOwnProperty(attr)) {
                    copy[attr] = arrays.clone(obj[attr]);
                }
            }
            return copy;
        }

        throw new Error('Unable to copy obj! Its type isn\'t supported.');
    };

    /**
     * bisector.
     */
    arrays.bisect = function (a, x) {
        var lo = 0;
        var hi = a.length;
        while (lo < hi) {
            var mid = lo + hi >>> 1;
            if (a[mid] < x) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        if (lo > 0 && (x - a[lo - 1]) < (a[lo] - x)) {
            lo -= 1;
        }
        return lo;
    };

    arrays.redirect = function (href) {
        window.location = href;
    };

    const validator = window.validator || {
        escape: _.identity,
        unescape: _.identity,
        isURL: _.stubFalse,
    };

    arrays.escape = function (value) {
        const string = _.trim(value);

        // skip URLs
        return validator.isURL(string) ? string : validator.escape(string);
    };

    arrays.unescape = function (value) {
        const string = _.trim(value);

        // skip URLs
        return validator.isURL(string) ? string : validator.unescape(string);
    };

    arrays.formatTick = function (d, scale, units) {
        var fixed = 0;

        // if domain is small (between 0â€“10), add decimals
        if (scale.domain()[0] >= 0 && scale.domain()[1] < 10) {
            fixed = 2;
        }

        var tick = arrays.orderOfMagnitude(d, fixed);

        if (units === '%') {
            tick += '%';
        }

        return tick;
    };

    arrays.getClickThroughURL = function (routePath_withoutFilter, filterObj, clickThroughView) {
        const [baseUrl, baseParams] = routePath_withoutFilter.split('?');
        const additionalParams = Qs.parse(baseParams);

        const url = clickThroughView ? `/${baseUrl.split('/')[1]}/${clickThroughView}` : baseUrl;
        const params = Qs.stringify(Object.assign({}, filterObj, additionalParams));

        return `${url}?${params}`;
    };
})(window.arrays, window.jQuery, window._, window.Qs);
