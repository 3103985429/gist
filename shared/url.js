const Qs = require('qs');
const _ = require('lodash');
const validator = require('validator');

module.exports = {
    constructedRoutePath,
    getRouteParams,
    extractHostname,
    getPreparedFilters,
};

function getRouteParams(routePath_base = '') {
    if (routePath_base.includes('?')) {
        const [routePath, routePath_queryParams] = routePath_base.split('?');
        return {
            routePath,
            routeParams: Qs.parse(routePath_queryParams),
        };
    }

    return {
        routePath: routePath_base,
        routeParams: {},
    };
}

function getPreparedFilters({ routeParams, filterObj, options }) {
    const queries = [
        'accessibility', 'aggregateBy', 'colFilter', 'chartBy', 'embed', 'groupBy', 'groupByDuration', 'groupByRange',
        'groupSize', 'horizontal', 'limit', 'mapBy', 'nested', 'normalize', 'page', 'preview', 'radius', 'revision',
        'searchCol', 'searchQ', 'secondaryCol', 'sortBy', 'sortDirection', 'stackBy', 'timeValue', 'xAxis', 'yAxis',
        'instanceId', 'objectIndex', 'viewType', 'queryFilters', 'segmentBy',
    ];
    const queryOptions = _.pick(options, queries);
    const queryObj = _.merge(routeParams, filterObj, queryOptions);
    const filteredQueryObj = _.omitBy(queryObj, (data) => data === '' || _.isNil(data));
    const preparedFilters = _.reduce(filteredQueryObj, (acc, param, key) => {
        const unescapeValue = (value) => _.isString(value) ? validator.unescape(value) : value;
        acc[key] = _.isArray(param) ? filteredQueryObj[key].map(unescapeValue) : param;

        return acc;
    }, {});

    return preparedFilters;
}

function constructedRoutePath(routePath_base, filterObj = {}, options = {}, isSharedPage) {
    const { routePath, routeParams } = getRouteParams(routePath_base);
    const preparedFilters = getPreparedFilters({ routeParams, filterObj, options });

    if (isSharedPage) {
        preparedFilters.queryFilters = true;
    }

    const mainPath = routePath.replace(/\/timeline$/, '/grouped-gallery');

    if (_.isEmpty(preparedFilters)) {
        return mainPath;
    }

    return `${mainPath}?${Qs.stringify(preparedFilters)}`;
}

function extractHostname(url) {
    let hostname = _.trim(url);

    if (hostname.includes('//')) {
        hostname = hostname.split('//')[1];
    }

    hostname = hostname.split('/')[0];
    hostname = hostname.split(':')[0];
    hostname = hostname.split('?')[0];

    return hostname.startsWith('www.') ? hostname.replace('www.', '') : hostname;
}
