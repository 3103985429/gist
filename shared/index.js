import windowArrays from 'arrays';
import fields from './fields';
import url from './url';
import markdown from './markdown';


export const arrays = {
    ...windowArrays,
    ...fields,
    ...url,
    ...markdown,
};
