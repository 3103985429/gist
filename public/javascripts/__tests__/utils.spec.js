import { getRouteParams, constructedRoutePath } from '../../../shared/url';


describe('Utils', () => {
    let arrays;

    beforeAll(() => {
        window.arrays = arrays = {};
        require('../utils');
        arrays.getRouteParams = getRouteParams;
        arrays.constructedRoutePath = constructedRoutePath;
    });

    beforeEach(() => {
        window.history.pushState({}, 'Test Title', '/');
    });

    describe('updateQuery', () => {
        let options;

        beforeEach(() => {
            options = {
                routePath_withoutFilter: '/some-chart?accessibility=true&someFilter=true',
            };
            jest.spyOn(window.history, 'replaceState');
        });

        afterEach(() => {
            options = {};
        });

        it('should update routePath_withoutFilter', () => {
            options.accessibility = false;
            arrays.updateQuery(options, 'accessibility', options.accessibility);

            expect(options.routePath_withoutFilter).toEqual('/some-chart?accessibility=false&someFilter=true');
        });

        it('should update history', () => {
            arrays.updateQuery(options, 'someFilter', true);

            expect(window.history.replaceState).toHaveBeenCalledWith({}, '', 'http://localhost/?someFilter=true');
        });
    });

    describe('getDetailsViewUrl', () => {
        it('should return url', () => {
            const options = {
                revision: true,
                embedded: false,
                array_source_key: 'array_source_key',
                filterObj: { Country: ['Poland'] },
            };

            expect(arrays.getDetailsViewUrl(options, 1, 0, 'map')).toMatchSnapshot();
        });
    });

    describe('changeRoutePath', () => {
        it('should return url', () => {
            window.history.pushState({}, 'Test Title', '/oldVizName?filterToOmit=1&filterToNotOmit=1');

            const filters = {
                filter: true,
                filter2: '2',
                filter3: false,
                filter4: 1,
            };

            const toOmit = ['filterToOmit'];

            expect(arrays.changeRoutePath('', 'vizName', filters, toOmit))
                .toEqual('http://localhost/vizName?filterToNotOmit=1&filter=true&filter2=2&filter3=false&filter4=1');
        });
    });

    describe('getApiRoute', () => {
        const options = {
            subdomain: 'test',
            isExternalAccess: false,
            environment: 'production',
            sharedPage: false,
            array_source_key: 'sourceKey/',
            sharedPageId: 1,
        };

        describe('when not sharedPage', () => {
            it('should return PROD link', () => {
                const link = arrays.getApiRoute(options, 'fake/link');
                const linkEnterprise = arrays.getApiRoute({ ...options, environment: 'enterprise' }, 'fake/link');

                expect(link).toBe('https://test.gist.info/json-api/v1/datasources/sourceKey/fake/link');
                expect(linkEnterprise).toBe('https://test.gist.info/json-api/v1/datasources/sourceKey/fake/link');
            });

            it('should return STAGING link', () => {
                const link = arrays.getApiRoute({ ...options, environment: 'staging' }, 'fake/link');

                expect(link).toBe('https://test.staging.gist.info/json-api/v1/datasources/sourceKey/fake/link');
            });

            it('should return DEV link', () => {
                const link = arrays.getApiRoute({ ...options, environment: 'development' }, 'fake/link');

                expect(link).toBe('http://test.local.arrays.co:9080/json-api/v1/datasources/sourceKey/fake/link');
            });
        });

        describe('when sharedPage', (sharedOptions = { ...options, sharedPage: true }) => {
            it('should return PROD link', () => {
                const link = arrays.getApiRoute(sharedOptions, 'fake/link');
                const linkEnterprise = arrays.getApiRoute({ ...sharedOptions, environment: 'enterprise' }, 'fake/link');

                expect(link).toBe('https://test.gist.info/json-api/v1/sharedpages/1/');
                expect(linkEnterprise).toBe('https://test.gist.info/json-api/v1/sharedpages/1/');
            });

            it('should return STAGING link', () => {
                const link = arrays.getApiRoute({ ...sharedOptions, environment: 'staging' }, 'fake/link');

                expect(link).toBe('https://test.staging.gist.info/json-api/v1/sharedpages/1/');
            });

            it('should return DEV link', () => {
                const link = arrays.getApiRoute({ ...sharedOptions, environment: 'development' }, 'fake/link');

                expect(link).toBe('http://test.local.arrays.co:9080/json-api/v1/sharedpages/1/');
            });

            it('should return direct link', () => {
                const link = arrays.getApiRoute({ ...sharedOptions, environment: 'development' }, 'fake/link', true);

                expect(link).toBe('http://test.local.arrays.co:9080/json-api/v1/datasources/sourceKey/fake/link');
            });
        });
    });
});

