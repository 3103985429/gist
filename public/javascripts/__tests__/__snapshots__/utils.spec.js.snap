// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils getDetailsViewUrl should return url 1`] = `"/array_source_key/1?Country%5B0%5D=Poland&revision=true&objectIndex=0&viewType=map"`;
